src/replay/service.zig

```diff
@@ -91,19 +96,1269 @@ fn advanceReplay(state: *ReplayState) !void {
+        fork_choice: *const sig.consensus.HeaviestSubtreeForkChoice,
+        epoch_slots_frozen_slots: *const EpochSlotsFrozenSlots,
+    ) DeadState {
+        _ = slot; // autofix
+        _ = duplicate_slots_tracker; // autofix
+        _ = duplicate_confirmed_slots; // autofix
+        _ = fork_choice; // autofix
+        _ = epoch_slots_frozen_slots; // autofix
+    }
+};
+
+pub const DuplicateState = struct {
+    duplicate_confirmed_hash: ?sig.core.Hash,
+    bank_status: BankStatus,
+
+    pub fn fromState(
+        logger: sig.trace.Logger,
+        slot: sig.core.Slot,
+        duplicate_confirmed_slots: *const DuplicateConfirmedSlots,
+        fork_choice: *const sig.consensus.HeaviestSubtreeForkChoice,
+        bank_status: BankStatus,
+    ) DuplicateState {
+        // We can only skip marking duplicate if this slot has already been
+        // duplicate confirmed, any weaker confirmation levels are not sufficient
+        // to skip marking the slot as duplicate.
+        const duplicate_confirmed_hash = getDuplicateConfirmedHashFromState(
+            logger,
+            slot,
+            duplicate_confirmed_slots,
+            fork_choice,
+            bank_status.bankHash(),
+        );
+        return .{
+            .duplicate_confirmed_hash = duplicate_confirmed_hash,
+            .bank_status = bank_status,
+        };
+    }
+};
+
+pub const EpochSlotsFrozenState = struct {
+    epoch_slots_frozen_hash: sig.core.Hash,
+    duplicate_confirmed_hash: ?sig.core.Hash,
+    bank_status: BankStatus,
+    is_popular_pruned: bool,
+
+    pub fn fromState(
+        logger: sig.trace.Logger,
+        slot: sig.core.Slot,
+        epoch_slots_frozen_hash: sig.core.Hash,
+        duplicate_confirmed_slots: *const DuplicateConfirmedSlots,
+        fork_choice: *const sig.consensus.HeaviestSubtreeForkChoice,
+        bank_status: BankStatus,
+        is_popular_pruned: bool,
+    ) EpochSlotsFrozenState {
+        const duplicate_confirmed_hash = getDuplicateConfirmedHashFromState(
+            logger,
+            slot,
+            duplicate_confirmed_slots,
+            fork_choice,
+            bank_status.bankHash(),
+        );
+        return .{
+            .epoch_slots_frozen_hash = epoch_slots_frozen_hash,
+            .duplicate_confirmed_hash = duplicate_confirmed_hash,
+            .bank_status = bank_status,
+            .is_popular_pruned = is_popular_pruned,
+        };
+    }
+};
+
+fn processAncestorHashesDuplicateSlots(
+    allocator: std.mem.Allocator,
+    logger: sig.trace.Logger,
+    pubkey: sig.core.Pubkey,
+    ancestor_duplicate_slots_receiver: *sig.sync.Channel(AncestorDuplicateSlotToRepair),
+    duplicate_confirmed_slots: *const DuplicateConfirmedSlots,
+    epoch_slots_frozen_slots: *EpochSlotsFrozenSlots,
+    progress: *const sig.consensus.ProgressMap,
+    fork_choice: *sig.consensus.HeaviestSubtreeForkChoice,
+    bank_forks_rwmux: *sig.sync.RwMux(sig.replay.trackers.SlotTracker),
```

Comment:
nit:
```suggestion
    slot_tracker_rwmux: *sig.sync.RwMux(sig.replay.trackers.SlotTracker),
```

Reply:
Would rather not start renaming these yet; I'm not convinced on `SlotTracker` as the final name for the lightweight bankforks, given that an extremely similar name is currently in use in vote tracker "SlotVoteTracker", and is often referred to as just "SlotTracker".